client/src/main/java/io/pravega/client/security/auth/JwtTokenProviderImpl.java

```diff
@@ -0,0 +1,236 @@
+     * @param streamName the name of the stream tied to the segment, for which a delegation token is to be obtained
+     */
+    public JwtTokenProviderImpl(String token, Controller controllerClient, String scopeName,
+                                String streamName) {
+        this(token, controllerClient, scopeName, streamName, ConfigurationOptionsExtractor.extractInt(
+                "pravega.client.auth.token-refresh.threshold",
+                "pravega_client_auth_token-refresh.threshold",
+                DEFAULT_REFRESH_THRESHOLD));
+    }
+
+    /**
+     * Constructs a new object with the specified {delegationToken}, {@code controllerClient}, {@code scopeName} and
+     * {@code streamName}.
+     *
+     * @param token the initial delegation token
+     * @param controllerClient the {@link Controller} client used for obtaining a delegation token from the Controller
+     * @param scopeName the name of the scope tied to the segment, for which a delegation token is to be obtained
+     * @param streamName the name of the stream tied to the segment, for which a delegation token is to be obtained
+     * @param tokenRefreshThreshold the time in seconds before expiry that should trigger a token refresh
+     */
+    public JwtTokenProviderImpl(String token, Controller controllerClient, String scopeName, String streamName,
+                                    int tokenRefreshThreshold) {
+        Exceptions.checkNotNullOrEmpty(token, "delegationToken");
+        Exceptions.checkNotNullOrEmpty(scopeName, "scopeName");
+        Preconditions.checkNotNull(controllerClient, "controllerClient is null");
+        Exceptions.checkNotNullOrEmpty(streamName, "streamName");
+
+        Long expTime = extractExpirationTime(token);
+        delegationToken.set(new DelegationToken(token, expTime));
+        this.scopeName = scopeName;
+        this.streamName = streamName;
+        this.controllerClient = controllerClient;
+        this.tokenRefreshThreshold = tokenRefreshThreshold;
+    }
+
+    /**
+     * Returns the delegation token. It returns existing delegation token if it is not close to expiry. If the token
+     * is close to expiry, it obtains a new delegation token and returns that one instead.
+     *
+     * @return String the delegation token JWT compact value
+     */
+    @Override
+    public String retrieveToken() {
+        if (isTokenNearingExpiry()) {
+            log.info("Token is nearing expiry");
+            return refreshToken();
+        } else {
+            return delegationToken.get().getValue();
+        }
+    }
+
+    @Override
+    public String refreshToken() {
+        log.trace("Entered refreshToken().");
+        DelegationToken existingToken = this.getCurrentToken();
+        if (isTokenBeingRefreshed.get()) {
+            log.debug("Token is already under refresh.");
+
+            // Wait for the token to finish refreshing
+            while (!isTokenBeingRefreshed.get()) {
```

Comment:
I don't understand this busy wait loop at all.

1. you can simplify it by looping on the condition of `isTokenBeingRefreshed.compareAndSet(false, true)` returning false. When it returns true, you got it and can reset the token.
2. You set `isTokenBeingRefreshed` to false inside `resetToken`. That makes your logic pretty hard to follow. You should be modifying it in the same level of the code, not in different methods.
3. Why the busy-wait loop and why sleep 200ms in between? `resetToken` does absolutely nothing interesting. It performs a `compareAndSet` which is an atomic operation and it has its own synchronization primitives. I don't understand why you need a busy loop at all here.

Reply:
Yes, it was quite ugly. I've now modified this part of the code significantly, which resolves these issues.